core/src/main/java/com/linecorp/armeria/internal/AnnotatedHttpServiceParamUtil.java

```diff
@@ -208,6 +216,127 @@ public static Object convertParameter(@Nullable String value,
+     *
+     * <p>Note that the name of the specified {@code nameRetrievalTarget} will be converted as
+     * {@link CaseFormat#LOWER_HYPHEN} that the string elements are separated with one hyphen({@code -})
+     * character. The value of the {@link Header} annotation will not be converted because it is clearly
+     * specified by a user.
+     */
+    @Nullable
+    public static String findHeaderName(AnnotatedElement element, Object nameRetrievalTarget) {
+        requireNonNull(element, "element");
+        requireNonNull(nameRetrievalTarget, "nameRetrievalTarget");
+        final Header header = element.getAnnotation(Header.class);
+        if (header == null) {
+            return null;
+        }
+
+        final String value = header.value();
+        if (DefaultValues.isSpecified(value)) {
+            checkArgument(!value.trim().isEmpty(), "value: %s (expected: not blank)", value);
+            return value;
+        }
+        return toHeaderName(getName(nameRetrievalTarget));
+    }
+
+    /**
+     * Returns whether the specified {@code element} is annotated with a {@link Param} or {@link Header}.
+     */
+    public static boolean hasParamOrHeader(AnnotatedElement element) {
+        return element.isAnnotationPresent(Param.class) ||
+               element.isAnnotationPresent(Header.class);
+    }
+
+    private static String getName(Object element) {
+        if (element instanceof Parameter) {
+            final Parameter parameter = (Parameter) element;
+            checkArgument(parameter.isNamePresent(),
+                          "cannot obtain the name of the parameter or field automatically. " +
+                          "Please check whether your compile option includes '-parameters'. " +
+                          "If not, you need to specify parameter and header names with @%s and @%s.",
+                          Param.class.getName(), Header.class.getName());
+            return parameter.getName();
+        }
+        if (element instanceof Field) {
+            return ((Field) element).getName();
+        }
+        throw new IllegalArgumentException("cannot find the name: " + element.getClass().getName());
+    }
+
+    @VisibleForTesting
+    static String toHeaderName(String name) {
+        requireNonNull(name, "name");
+        checkArgument(!name.trim().isEmpty(), "name: %s (expected: not blank)", name);
+
+        if (name.equals(name.toUpperCase())) {
+            return CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_HYPHEN, name);
+        }
+        if (name.equals(name.toLowerCase())) {
+            return CaseFormat.LOWER_UNDERSCORE.to(CaseFormat.LOWER_HYPHEN, name);
+        }
+        // Ensure that the name does not contain '_'.
+        // If it contains '_', we give up to make it lower hyphen case. Just converting it to lower case.
+        if (name.indexOf('_') >= 0) {
+            return name.toLowerCase();
```

Comment:
- Please use `Ascii.toLowerCase()` for all `String.toLowerCase()` calls.
- The lowercased `name` could be extracted into a local variable because it appears twice.